File: MNA_Methods/MATLAB/1-Piece by Piece MNA/Load_netlist.m
%---------------------------------------------------------%

% Load_netlist.m
% 
% Author: Rasul Choupanzadeh
% Date: 05/12/2022

%---------------------------------------------------------%




function [Ra,Rb,L,C,Num_branch]=Load_netlist(sbck)
global netlist ;
file='netlist_passive.sp';
fileID = fopen(file);
netlist = textscan(fileID,'%s %s %s %s %s %s %s %s %s');
fclose(fileID);
clear fileID file ans;
%-------------------------------Split the sub-circuits------------------------------------
subckt_identifier=[];
nt=1;
for i=1:length(netlist{1})
    s = netlist{1}{i};
    st= '.ends';
    tf = strcmp(s,st);                          % Compares the strings with '.ends'
    if tf==1
        subckt_identifier(nt)=i+1;              % Stores the row numbers for the begining of each sub_circuit; which is occures after each '.ends'
        nt=nt+1;
    end  
end

% Extracting parameters of sub-circuit to calculate MNA matrices
Num_sbckt=length(subckt_identifier)-1; 
Num_branch=0;
for i=subckt_identifier(sbck):subckt_identifier(sbck+1)-1
    s = netlist{2}{i};
    st= 'Branch';
    tf = strcmp(s,st);                      
    if tf==1
        Num_branch=Num_branch+1;              
    end  
end
Ra=zeros(1,Num_branch);
L=zeros(1,Num_branch);
Rb=zeros(1,Num_branch);
C=zeros(1,Num_branch);
nt1=1;
nt2=1;
nt3=1;
nt4=1;
for i=subckt_identifier(sbck):subckt_identifier(sbck+1)-1
    s = netlist{1}{i};
    switch(s(1))
        case{'R'}
            switch(s(2))
                case{'a'}
                    Ra(nt1)= str2double(netlist{4}{i});
                    nt1=nt1+1;
                case{'b'}
                    Rb(nt2)= str2double(netlist{4}{i});
                    nt2=nt2+1;
            end
        case{'L'}
            L(nt3)= str2double(netlist{4}{i});
            nt3=nt3+1;
        case{'C'}
            C(nt4)= str2double(netlist{4}{i});
            nt4=nt4+1;
    end
end
end

File: MNA_Methods/Python/1-Piece by Piece MNA/Load_netlist.py

''' Load_netlist.py

Author: Rasul Choupanzadeh
Date: 05/12/2022

'''



import numpy as np

# Load and read the netlist line by line    
netlist = open('netlist.sp').readlines()                

# calculate the number of sub-circuits and ports
Num_sbck = 0;
sbck_ident = [];
for i in range(len(netlist)):
    if netlist[i][0]=='X':
        Num_sbck = Num_sbck+1
        Num_port = int(netlist[i][2])
    if netlist[i][0:4]=='.sub':
        sbck_ident.append(i)        
del sbck_ident[0]                       ## to delete a row from list==> del matrix[i,:]          to delete a row from a numpy array==> np.delete(array name, row/column number, axis=0/1 for row/column)
sbck_ident.append(len(netlist))


# define function to call sub-circuits
def Load_sbck(sbck):
    Rabr = []; Lbr = []; Rbbr = []; Cbr = [];
    Num_branch = 0;
    j = sbck-1;
    for i in range(sbck_ident[j], sbck_ident[j+1]):
        
        if netlist[i][0:8]=='* Branch':
            Num_branch = Num_branch+1;
            
        if netlist[i][0:2]=='Ra':
            sentence = netlist[i];
            for t in sentence.split():
                try:
                    R = float(t)
                except ValueError:
                    pass
            Rabr.append(float(R))
            
        if netlist[i][0:2]=='Rb':
            sentence = netlist[i];
            for t in sentence.split():
                try:
                    R = float(t)
                except ValueError:
                    pass
            Rbbr.append(float(R))  
            
        if netlist[i][0:2]=='Lb':
            sentence = netlist[i];
            for t in sentence.split():
                try:
                    l = float(t)
                except ValueError:
                    pass
            Lbr.append(float(l))   
        if netlist[i][0:2]=='Cb':
            sentence = netlist[i];
            for t in sentence.split():
                try:
                    c = float(t)
                except ValueError:
                    pass
            Cbr.append(float(c))              
    Ra = np.array(Rabr);
    Rb = np.array(Rbbr);
    L = np.array(Lbr);
    C = np.array(Cbr);
    return Num_branch, Ra, L, Rb, C



